### DISCUSSION TITLE
What is a correct way to refresh access tokens

### DISCUSSION THREAD

**Discussion by adanielyan** [remix-run/remix#4592] [Q&A]
I am new to remix, and so far I love this framework.

However I've been struggling with one issue for a few days and I am now stuck. I have Directus as my headless CMS and authentication service for my users. The way auth works is you get an access token and a refresh token. The refresh token is used to get a new access token when the old one expires. The problem is that I don't know where to put my access token refresh code.

If I put it in the loader of ```root.tsx``` it works fine for the initial load, but then if I start navigating from one route to another and a token refresh is needed at some point, the code in ```root.tsx``` doesn't fire. So I end up with failed requests in my loaders. If I place that code in every route that requires authentication, then I have code racing on initial load when there are nested routes that have the token refresh code in their loaders. I cannot have the token refresh only in the leaf routes, as I have parent routes that also need the user object, for example to render a user menu in the page navigation.

I should say that I user Remix Server and would prefer not switching to Express unless I absolutely have to.

I don't know how to proceed and this issue is just driving me crazy. Have anyone else had to deal with this problem? I would appreciate any advice.

**Comment by kiliman** [comment]
Here's some pseudocode that should help. The "trick" is to verify the token, and if it's expired, get a new token and redirect back to the same page. You should now have the correct token.

```ts
export const loader = async ({request}) => {
  const token = await verifyToken(request)
  // the token will be valid here because it either hasn't expired
  // or we refreshed the token and threw a redirect with new token

  //... do what you need to do with the token
}
 
async function verifyToken(request: Request) {
  let token = await getToken(request)
  if (isExpired(token)) {
    token = await refreshToken(request)
    // token is now valid, so redirect back to same page
    // and Remix will resend the request with the new token
    // we throw here so we don't have to check return in loader
    throw redirect(request.url, { headers: { 'set-coookie': tokenCookie }})
  }
  // token should be valid here, so return it
  return token
}
```

**Reply by seanlennaerts** [reply]
Thanks! This trick works nicely for loaders. For actions you have to be careful to not redirect (see https://sergiodxa.com/articles/working-with-refresh-tokens-in-remix for more details).

The last edge case to consider is making client side fetches (ie useSubmit) to resource route loaders (ie a search input showing results dynamically). Not sure what the trick is here:
* doing a check on `request.method === 'GET'` is too broad since this will include our search resource route (GET `/search?q=foo`). Redirecting here will actually try to load `/search` as a page route (probably a blank page).
* we need to somehow differentiate route loaders from resource route loaders

**Reply by myeljoud** [reply]
@seanlennaerts you've raised up a good point about resource routes and I'm actually having that problem, @sergiodxa I've read your [article](https://sergiodxa.com/articles/working-with-refresh-tokens-in-remix) but it does not seem to account to this case, is there a workaround for this?

**Reply by sergiodxa** [reply]
The best approach in my experience so far is to never store the access token, instead use the refresh token to get a new access token on every request

Specially if the access token is short lived 

If the access token is long lived keep it around, and check from time to time the expiration to refresh it, ideally in background

**Reply by adanielyan** [reply]
If you have the single fetch enabled this problem is easily solved by putting your token refresh logic in a root loader. Since it will be called on every request you can effectively use it as middleware.

**Reply by sergiodxa** [reply]
In some cases loaders can be called individually, even with Single Fetch, so while in most cases it may be like a middleware, but it's not safe to use as one

**Reply by adanielyan** [reply]
That's interesting! Can you please elaborate? In what cases loaders can be called individually?

**Reply by sergiodxa** [reply]
Every route with shouldRevalidate or clientLoader will be fetched individually, even if the shouldRevalidate returns true or if clientLoader calls the serverLoader immediately.

So If you have `app/root` and `app/routes/_index` and your index route has both a loader and clientLoader, every client-side navigation will cause two HTTP requests, so the `app/root` won't work as a middleware for the index route.

Same thing if you have a shouldRevalidate in `app/root`, now that one will cause a separate request, so even if all of your other routes are fetched in the same request, the `app/root` won't, and many people use shouldRevalidate in `app/root` to prevent it from running in other scenarios.

**Reply by adanielyan** [reply]
Thanks for explaining!

**Reply by gustavopch** [reply]
Just want to point out that middleware support is right around the corner btw.

**Comment by adanielyan** [comment]
Thank you for your help, @kiliman! 

Assuming I have nested routes and in the loaders of each route I need to make an authenticated request to my API (and therefore a valid token), should I call the verifyToken() function in each of them or only in the root?

UPDATE: If I place this only in root's loader than the verifyToken gets called only on initial page load, but not when navigating from page to page. If I place it in loaders of other routes then this function gets called multiple times on initial route which eventually leads race condition and multiple token refreshes. That's exactly what my issue is.

**Reply by adanielyan** [reply]
It seems there's just no way to guarantee having the access tokens from the most recent request to refresh API in the session cookie if there's more than one loader in the nested route hierarchy that can call that API.

I wonder how others solve this problem. Refreshing access tokens is a very common procedure and I don't think I am the only one who had to deal with it in Remix. Please share your experience and solution if you found any!

**Reply by kiliman** [reply]
Unfortunately since Remix calls loaders in parallel, if you have multiple routes that need to refresh the token, you'll probably need some kind of shared cache. Unfortunately even that may not be enough depending on what host you're using. For example, serverless may run loaders in different processes or even machines.

Does refreshing a token automatically invalidate the other tokens, even if they haven't expired yet?

**Reply by adanielyan** [reply]
Yes, unfortunately refreshing the token does invalidate all other tokens for that user. 

What I think I need is being able to run a piece of code that checks the token (and refreshes it if needed) only once per each transition. If there was a way to run a hook or function only once per transition this problem would be solved very easily.

**Reply by kiliman** [reply]
A pre-request hook is coming, unfortunately that doesn't help you today.

**Reply by adanielyan** [reply]
Will it run only once per transition or per request (i.e. before each loader or action)?

**Reply by kiliman** [reply]
It runs once per transition (logical request). Useful for authentication, etc.

This will be in React Router first, but will be part of Remix.
https://github.com/remix-run/react-router/discussions/9564

**Reply by sergiodxa** [reply]
Even with a middleware you will keep having this issue, because on client-side navigation Remix will only call the loaders that changed and each loader will have an individual request.

This means if two loaders changed then you'll have two different HTTP requests made to your server and both will run the middleware and refresh the token, this will be worst in serverless or edge platforms where each request is also a separate process so you can't even share an in-memory cache between them.

**Reply by sergiodxa** [reply]
In this scenario, it's better to save the access token in a DB/another storage you can query at any time, and then run a background job to refresh the tokens. Then in your loaders/actions you can just get the access token from this storage and use it without caring about refreshing the token.

**Reply by adanielyan** [reply]
> This means if two loaders changed then you'll have two different HTTP requests made to your server and both will run the middleware and refresh the token...

That was exactly my concern and the reason I asked the question above about one execution per transition. So what you are suggesting is that the beforeLoader (or middleware if you wish) will be called per request and not per route transition? If that's the case, then indeed the middleware will not help, as it only eliminates the need to write the same code in multiple places, but not the number of times it's called. 

Just to be clear, isn't there any place in remix to have a function that will only run once per route transition? Sorry for my confusion, maybe I don't understand the differences between transitions and requests, but Michael's answer above suggests that there will be only one call of the beforeLoader hook per transition.

**Reply by kiliman** [reply]
I'm pretty sure the proposed `beforeLoader` will handle this case. Otherwise, you might as well not have it if it also runs in parallel.

My understanding is that:

* React Router determines what loaders to call (full or partial route)
* Runs all `beforeLoaders` in current route tree (root to leaf) **serially**
* Runs route `loaders` in **parallel**

So if you have a `beforeLoader` in _root_, and you're on _/parent/child-A_ and navigate to _/parent/child-B_

* RR will only call _child-B_ loader, since _root_ and _parent_ loaders are common ancestor
* RR will still call _root_ `beforeLoader`
* Then RR will call _child-B_ `loader`

**Reply by sergiodxa** [reply]
> as it only eliminates the need to write the same code in multiple places, but not the number of times it's called.

Exactly, this is what middleware do, let's you avoid repeating code, every request will run middleware once, on a document request that means multiple loaders will share the same middleware but in data requests is one request per loader.

> Just to be clear, isn't there any place in remix to have a function that will only run once per route transition? Sorry for my confusion, maybe I don't understand the differences between transitions and requests

There's no way and there can't be a way, remember loaders runs only server-side, and loaders are GET HTTP request, when you do a document request (this is a request to get an HTML response) Remix receives a single HTTP request, calls all loaders, gets the data and send a response.

With middlewares/beforeLoader Remix will first call middleware and then all loaders.

Once the app hydrates client-side, it works as an SPA, so when the user goes from `/dashboard` to `/profile/edit` (for example) Remix needs the data of three loaders

1. the root route loader
2. the profile route loader
3. the edit route loader

Because the root route didn't changed, Remix avoid sending a request (Remix will request again if the search params change or after an action runs), so Remix will need to fetch the data of only the profile and edit routes.

This is a client-side transition or navigation, here Remix will send a new HTTP request for profile and another completely different request for edit, imagine it like doing this:

```ts
let [profileReponse, editResponse] = await Promise.all([
  fetch(profileUrl),
  fetch(editUrl)
])
```

Because each loader receives a request even with middlewars/beforeLoader there's no way to run the middleware once for both request because middlewares run server-side and server-side there are two completely separate requests.

This is aggravated when using serverless or edge deployment targets, like Cloudflare, Netlify, Vercel, AWS Lambda, etc. because here each request the server receives create a new process running your app, so they are not even on the same process, even worst they may be on completely different machines, so not even the file system could be shared.

**Reply by sergiodxa** [reply]
> I'm pretty sure the proposed `beforeLoader` will handle this case. Otherwise, you might as well not have it if it also runs in parallel.
> 
> My understanding is that:
> 
> * React Router determines what loaders to call (full or partial route)
> * Runs all `beforeLoaders` in current route tree (root to leaf) **serially**
> * Runs route `loaders` in **parallel**
> 
> So if you have a `beforeLoader` in _root_, and you're on _/parent/child-A_ and navigate to _/parent/child-B_
> 
> * RR will only call _child-B_ loader, since _root_ and _parent_ loaders are common ancestor
> * RR will still call _root_ `beforeLoader`
> * Then RR will call _child-B_ `loader`

The problem here is that how would you share data between loaders on client-side navigation? If Remix choose to run beforeLoaders client-side so it runs before each HTTP request it will have to use a POST to call your loaders in order to send the data, and you can't use server-side code in beforeLoaders anymore. If Remix runs beforeLoaders server-side then each loader will have a new requests.

Middlewares are to avoid duplicating code on each route handler, that's it, you need to add auth protections? Do it on the middleware and it runs before your loaders without thinking about it.

If Remix wanted to let you share state between data request to the loaders it would have to merge all data requests in a single http request so they can all run together.

**Reply by adanielyan** [reply]
> In this scenario, it's better to save the access token in a DB/another storage you can query at any time, and then run a background job to refresh the tokens. Then in your loaders/actions you can just get the access token from this storage and use it without caring about refreshing the token.

My understanding of refresh tokens is that they are meant to expire if user doesn't use the application for a predefined amount of time (e.g. one week). Otherwise they get refreshed together with access tokens. If I constantly update the access token (and as a result the refresh token since they cannot be refreshed separately) in the backend without action from a user, then I can just configure my CMS to make the tokens not expireable in the first place.

**Reply by adanielyan** [reply]
Thank you for detailed explanations, they are very helpful. A couple of follow up questions.

> I'm pretty sure the proposed `beforeLoader` will handle this case. Otherwise, you might as well not have it if it also runs in parallel.
 
If I understand correctly all I will need is to have my token refresh code in the root's `beforeLoader()` function and not anywhere else.

> The problem here is that how would you share data between loaders on client-side navigation?

I don't need to share data, I only need to update refresh token if the old one expired and send it back in a `redirect(request.url, {headers: ...})` which will hopefully stop the loaders from executing during the current request with the outdated token and will force them to execute on the new request that will be triggered by redirect, this time with a refreshed token.

> If Remix choose to run beforeLoaders client-side...

If it's a middleware, why would it run in on the client-side??

> There's no way and there can't be a way, remember loaders runs only server-side, and loaders are GET HTTP request, when you do a document request (this is a request to get an HTML response) Remix receives a single HTTP request, calls all loaders, gets the data and send a response.

OK, so in the current state of remix this is not possible. This is good to know.

> With middlewares/beforeLoader Remix will first call middleware and then all loaders.

Therefore when the middleware/beforeLoader is implemented it will be possible to refresh the token only once per transition, assuming the refresh code is only present in the root's `beforeLoader()` function.  And also assuming the `beforeLoader()` function in my `root.tsx` runs (and runs only once) when I transition from `/parent/child-A` to `/parent/child-B` even if the loader of `root.tsx` doesn't have to run. Is that correct?

**Reply by kiliman** [reply]
Yes, this is my understanding. Remember, this feature will first be implemented in React Router which will run client-side. But even when it makes it to Remix, the same process will occur. All `beforeLoaders` will run serially before the route `loaders` which run in parallel. Only the inital document request will run these directly on the server. The rest will be fetch calls from the client just like Remix does today.

From the proposal, you will also be able to share data from your `beforeLoaders` with your route `loaders` via context.

**Reply by adanielyan** [reply]
Thank you for confirming. Even if it is called from the client the actual code will still be executed on the server, right?

I think the beforeLoader will be a solution I'm looking for. In the meantime I'll just stick with tokens that only expire when user logs out explicitly since my app is currently in development anyway. Hopefully this feature will make it into a release within a couple of months.

**Reply by kiliman** [reply]
I assume so, however the initial version in RR will be all client-side. Although your `beforeLoader` can make fetch calls as needed.

**Reply by barbinbrad** [reply]
Doesn't the `beforeLoader` just kick the can up the road? Isn't the problem still that there are multiple `beforeLoader` calls, but each of them would invalidate the other?

What if you stored the refresh token in an HTTP-only cookie, but gave the client access to the access token, so that it could check whether it was expired in a setInterval function -- and then if it was close to expiring, you could hit a remix resource route that does the refreshing. 

The problem is similar on the client-side if you have multiple tabs open running the same check. But here you use localStorage mutex to prevent all the tabs from making the same request (assuming the access token that was being checked was being shared in localStorage).

**Reply by adanielyan** [reply]
I tried calling a fetcher in useEffect in my root.tsx when the transition becomes "submitting", but the root component doesn't get re-rendered when navigating from one child route to another.  Which means that I will have to call a fetcher to refresh endpoint on a client sides of my leaf routes, which is not much different from calling it in loaders. 

Calling a fetcher in setInterval sounds way too hacky to me.

**Reply by barbinbrad** [reply]
Totally understanding your perspective. I guess the idea I was trying to advocate for is that you wouldn't need to check on each request IF you made sure you checked _before_ any request is made. Here's an example where you'd run some code like this in your root if the user was authenticated:

```js
React.useEffect(() => {
   sessionService.ensureSession();
}, []);
```

And here's what `ensureSession()` would look like: https://github.com/gravitational/webapps/blob/4da005d774cff0820b50439541a3d6ee1f89d0b3/packages/teleport/src/services/websession/websession.ts

**Reply by adanielyan** [reply]
Thank you for sharing a code example. It's an interesting approach. While it's not really a remix way, it may be the only way for now.

**Reply by barbinbrad** [reply]
Definitely! Parallelized requests are great for performance, but make shared auth state trickier. I'm looking forward to see where this discussion goes over time.
